### PR TITLE
Make Scylla build with no warnings

### PR DESCRIPTION
**Pull request by haaawk**
This PR fixes all the compilation warnings so that the projects builds with no warnings at all.

Test: unit(dev)

### CODE REVIEW

File: database.cc
Comment (nyh) [inline database.cc:430]: I don't like this const_cast. Why do we need this?
Unless I'm missing something, io_priority_class only holds just an _id, and that's it - not the shares. Since modifying the shares doesn't modify this id, the update_shares should work fine on a const reference. Maybe we forgot to allow this in Seastar?

**Review by nyh — CHANGES_REQUESTED** [review]
I left one request (to avoid using const_cast and try to look for a better fix), but in general all of your patches deal with the changing inferfaces of the IO scheduler so let's have this series reviewed also by someone who's a bigger expert than myself in the current IO scheduler implementation.

File: database.cc
Comment (haaawk) [reply database.cc:431]: I'm not a big fan of const_cast either but that was the best solution I could come up with without changing Seastar API.

I've sent out "[PATCH] io_priority_class: Make update_shares const" to Seastar mailing list and added a FIXME here to remove const_cast once Seastar API is changed.

Maybe we could merge this and remove the const_cast once Seastar API is changed? If it's changed at all because the addition of const to io_priority_class::update_shares is a bit controversial. I think having a clean build with no warnings while  we're discussing Seastar API is valuable - what do you think @nyh ?

**Comment by haaawk** [comment]
v2:
I added FIXME to remove const_cast when/if io_priority_class::update_shares becomes marked as `const`.

**Comment by haaawk** [comment]
@xemul Could you please have a look, as an author of changes in IO Scheduler API in Seastar? (on @nyh request for more knowledgable reviewer)

File: database.cc
Comment (nyh) [reply database.cc:431]: 1. Why is adding the const in Seastar controversial? Who was against it?
2. This patch is optional - it just cleans a warning, doesn't fix any bug or feature - so why hurry with it? If we think that fixing Seastar is better than the ugly "const_cast", then let's wait a few more hours for it. We can commit the two other patches separately from this one (but I'd really appreciate a review from someone who's a bigger expert in this code than me).

Comment (haaawk) [reply database.cc:431]: 1. No one is against it at the moment but I can imagine a reasoning that setting shares for io_priority_class actually changes this class because now it gets more/less priority. The alternarive narration that io_priority_class is just an ID is also valid. I'm just saying that there are more than one sane interpretation.
2. I really don't have any problems with few hours :). I'm just not sure how long it takes to push seastar changes and seastar update in Scylla. I somehow assumed it's much longer than just changing Scylla alone. Maybe I had wrong impression.

**Review by xemul — APPROVED** [review]
I don't mind, but have one suggestion for the 2nd patch.

**Comment by haaawk** [comment]
I removed const_cast. @nyh Could you have a look please?

**Comment by haaawk** [comment]
Actually I also have to address @xemul comment.

**Comment by haaawk** [comment]
I addressed @xemul comment and passed opts to verify_seastar_io_scheduler

File: main.cc
Comment (haaawk) [reply main.cc]: done.

File: database.cc
Comment (haaawk) [reply database.cc]: Removed the const_cast

File: database.cc
Comment (nyh) [inline database.cc:432]: You forgot to remove this line

**Review by nyh — CHANGES_REQUESTED** [review]
Looks good except one FIXME comment from the previous version which you forgot to remove...

File: database.cc
Comment (haaawk) [reply database.cc:432]: Right. Give me a sec.

Comment (haaawk) [reply database.cc]: removed

**Comment by haaawk** [comment]
Removed outdated FIXME.

**Review by nyh — APPROVED** [review]
Thanks. I'll commit this but want to wait for "next" will promote to master (we already have more than a hundred unpromoted patches).

**Comment by haaawk** [comment]
Rebased on top of master

**Comment by haaawk** [comment]
It seems that seaster is still not updated in Scylla @nyh :(
```
database.cc:432:24: error: 'this' argument to member function 'update_shares' has type 'const ::io_priority_class', but function is not marked const
    _inflight_update = _io_priority.update_shares(uint32_t(shares));
                       ^~~~~~~~~~~~
/home/haaawk/haaawk/scylla-fresh/seastar/include/seastar/core/io_priority_class.hh:54:14: note: 'update_shares' declared here
    future<> update_shares(uint32_t shares);
             ^
1 error generated.
```

**Comment by haaawk** [comment]
This is the Seastar commit that this change depends on https://github.com/scylladb/seastar/commit/e5dd0bc37b66c1ee60ef0a9dc93e165e23a7afc3

**Comment by nyh** [comment]
It seems that for the last two weeks, every time I tried to update Seastar in Scylla, @avikivity  reverted the update - first it was because of gdb problems (those were already fixed) and later because of "strange errors with boost::condition_variable". Avi, did you ever get to the bottom of that condition_variable thing? We can't continue to be unable to update Seastar :-( If you tell me which test broke, I can try to debug it myself.

@haaawk, why not split the single patch that relies on this Seastar change to a separate PR - so I can commit all the rest?

**Comment by avikivity** [comment]
> It seems that for the last two weeks, every time I tried to update Seastar in Scylla, @avikivity reverted the update - first it was because of gdb problems (those were already fixed) and later because of "strange errors with boost::condition_variable". Avi, did you ever get to the bottom of that condition_variable thing? We can't continue to be unable to update Seastar :-( If you tell me which test broke, I can try to debug it myself.
> 
> @haaawk, why not split the single patch that relies on this Seastar change to a separate PR - so I can commit all the rest?


There's a week-old patch waiting for review.

**Comment by haaawk** [comment]
I've extracted two commits that don't depend on Seastar change to separate PR -> https://github.com/scylladb/scylla/pull/8850
This PR keeps only a single commit that needs Seastar update.

**Comment by nyh** [comment]
> > It seems that for the last two weeks, every time I tried to update Seastar in Scylla, @avikivity reverted the update - first it was because of gdb problems (those were already fixed) and later because of "strange errors with boost::condition_variable". Avi, did you ever get to the bottom of that condition_variable thing? We can't continue to be unable to update Seastar :-( If you tell me which test broke, I can try to debug it myself.
> > @haaawk, why not split the single patch that relies on this Seastar change to a separate PR - so I can commit all the rest?
> 
> There's a week-old patch waiting for review.

Thanks Avi, I missed that patch... I'll commit it now, then try to update Seastar in Scylla and finally commit @haaawk 's last patch.

**Pull request closed without merging**
